### PR TITLE
Feature/col 168 cms component

### PR DESCRIPTION
**Pull request by darrellme**
Added the CMS Component and build functionality on top of the feature/COL-168 feature branch

### CODE REVIEW

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  name: 'Mock Product Carousel',
+  strategy: 'test-strategy-1',
+  container: 'false',
+};
+
+const MockCmsMerchandisingCarouselComponent = <CmsComponentData<any>>{
+  data$: of(mockComponentData),
+};
+
+// class MockProductService {
```
Comment (bgambocjaviniar) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts]: remove commented code

File: projects/storefrontapp-e2e-cypress/cypress/integration/smoke/merchandising-carousel.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+context('Merchandising Carousel', () => {
```
Comment (bgambocjaviniar) [inline projects/storefrontapp-e2e-cypress/cypress/integration/smoke/merchandising-carousel.e2e-spec.ts]: this shouldn't be in the smoke test. It would be in the regression folder.
This is like a unit test, and not an e2e test, where we test a user's flow

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    ),
+    map(merchandisingProducts => merchandisingProducts.products),
+    map(products => products.map(product => of(product)))
+  );
+
+  constructor(
+    protected componentData: CmsComponentData<
+      CmsMerchandisingCarouselComponent
+    >,
+    protected strategyConnector: StrategyConnector
```
Comment (bgambocjaviniar) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: instead of injecting the connector directly here, let's create a facade layer to align what we do in Spartacus even though we're not using the store. 

Basically, I think we should inject the connector in that facade layer and delegate the call here

File: projects/cds/public_api.ts [hunk 4]
```diff
@@ -2,6 +2,8 @@
  * Public API Surface for CDS
  */
 export * from './src/cds-models/index';
+export * from './src/cds.module';
+export * from './src/cms-components/index';
```
Comment (bgambocjaviniar) [inline projects/cds/public_api.ts]: we shouldn't expose things to the public until we're 100% sure. We should include this in the final merge to have room for improvements as if we include it now, we can't change function names and such.

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<cx-carousel
+  [items]="items$ | async"
+  [title]="title$ | async"
+  [template]="merchandisingCarouselItem"
```
Comment (bgambocjaviniar) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html]: we should keep the naming convention consistent through carousels, therefore I believe it should be named carouselItem

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+export class MerchandisingCarouselComponent {
+  private componentData$: Observable<
+    CmsMerchandisingCarouselComponent
+  > = this.componentData.data$.pipe(filter(Boolean));
+
+  title$: Observable<string> = this.componentData$.pipe(
+    map(data => data.title)
+  );
+
+  size$: Observable<string> = this.componentData$.pipe(
```
Comment (bgambocjaviniar) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: I know we need size$, backgroundColor$, and textColor$ as merch carousel supports these additional properties, but we shouldn't include it in this PR until make changes to the current carousel implementation to support these changes.

File: projects/storefrontstyles/scss/_components.scss [hunk 7]
```diff
@@ -31,14 +32,15 @@ $selectors: mergeAll(
...
-    $pwa-components-whitelist
+    $pwa-components-whitelist,
+    $cds-components-whitelist
   )
 );
 
 @each $selector in $selectors {
   #{$selector} {
     // skip selectors if they're added to the $skipComponentStyles list
-    @if (index($skipComponentStyles, $selector) == null) {
```
Comment (bgambocjaviniar) [inline projects/storefrontstyles/scss/_components.scss]: please revert this change

File: projects/storefrontstyles/scss/_components.scss [hunk 8]
```diff
@@ -49,7 +51,7 @@ $selectors: mergeAll(
 // add body specific selectors
 body {
   @each $selector in $selectors {
-    @if (index($skipComponentStyles, $selector) == null) {
```
Comment (bgambocjaviniar) [inline projects/storefrontstyles/scss/_components.scss]: same as above

File: projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss [hunk 9]
```diff
@@ -1,4 +1,5 @@
-%cx-product-carousel {
+%cx-product-carousel,
+%cx-merchandising-carousel {
```
Comment (bgambocjaviniar) [inline projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss]: Please create a _merchandising-carousel.scss, and extend the cx-product-carousel if needed, so we can place stylings accordingly to each 'component'

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<cx-carousel
```
Comment (bgambocjaviniar) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html]: where's the metadata for the carousel level as we discussed before?

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  [items]="items$ | async"
+  [title]="title$ | async"
+  [template]="merchandisingCarouselItem"
+  itemWidth="285px"
+>
+</cx-carousel>
+
+<ng-template #merchandisingCarouselItem let-item="item">
+  <a tabindex="0" [routerLink]="{ cxRoute: 'product', params: item } | cxUrl">
```
Comment (bgambocjaviniar) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html]: where's the metadata for the product level?

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  );
+
+  items$: Observable<Observable<Product>[]> = this.componentData$.pipe(
+    map(data => data.strategy),
+    distinctUntilChanged(),
+    switchMap(strategyId =>
+      this.strategyConnector.loadProductsForStrategy(strategyId)
+    ),
+    map(merchandisingProducts => merchandisingProducts.products),
+    map(products => products.map(product => of(product)))
```
Comment (znikola) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: I think this line can be removed. This will flatten the return type (line 35) to just `Observable<Product[]>`

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 6 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: I've removed them for now, however, size relates more to how many products are returned from the strategy service. We have another ticket to include that functionality in the next sprint. One question though, if we move this stuff from the component, it still shows in the configuration of the CMS component in smartedit....is that ok?

File: projects/storefrontstyles/scss/_components.scss (hunk 7 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/_components.scss]: It looks like this was reverted in one of the later commits, as I see this line in the feature/COL-168-cms-compent branch https://github.com/SAP/cloud-commerce-spartacus-storefront/blob/feature/COL-168-cms-component/projects/storefrontstyles/scss/_components.scss

File: projects/storefrontstyles/scss/_components.scss (hunk 8 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/_components.scss]: As above

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html (hunk 10 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html]: This will be done in ticket https://jira.hybris.com/browse/COL-164

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html (hunk 11 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html]: This will be done in ticket https://jira.hybris.com/browse/COL-164, which will be in the next sprint starting next week

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 12 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: This can't be removed. Its something Nikola mentioned also, however, if we remove it and only have an Observable<Product[]> then we get console errors due to the AsyncPipe on the cx-carousel. Also, the OCC product carousel does this it in the same way, which is where we based it off of

File: projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss (hunk 9 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss]: Done

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html (hunk 5 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.html]: Done

File: projects/cds/public_api.ts (hunk 4 above)
Comment (darrellme) [reply projects/cds/public_api.ts]: We need to export it to be able to test it in the storefront application, no?

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 3 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: Ok, I've created a CdsMerchandisingProductService that is used instead. Is this what you meant?

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts (hunk 1 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts]: done

File: projects/storefrontapp-e2e-cypress/cypress/integration/smoke/merchandising-carousel.e2e-spec.ts (hunk 2 above)
Comment (darrellme) [reply projects/storefrontapp-e2e-cypress/cypress/integration/smoke/merchandising-carousel.e2e-spec.ts]: I've moved the test to the regression area. We definitely need a test like this as its more of an integration test, rather than a unit test, to ensure that the carousel is using our strategy service and the payload is being translated correctly. Whether we need an actual e2e test, I'm not convinced, as the carousel will display regardless of metrics (user behaviour)

Comment (bgambocjaviniar) [reply projects/storefrontapp-e2e-cypress/cypress/integration/smoke/merchandising-carousel.e2e-spec.ts]: as discussed in the meeting, we will most likely drop this e2e test because we would need a dedicated cds instance for Spartacus.

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 6 above)
Comment (bgambocjaviniar) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: I think it would be best to include these 'configs' in the next ticket as well as it is not used as of yet. Like I mentioned, I know it is necessary since these config exists for the merch carousel and used in smartedit, but the current carousel does not support these inputs as of yet.

File: projects/storefrontapp-e2e-cypress/cypress/integration/smoke/merchandising-carousel.e2e-spec.ts (hunk 2 above)
Comment (darrellme) [reply projects/storefrontapp-e2e-cypress/cypress/integration/smoke/merchandising-carousel.e2e-spec.ts]: I've removed this

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 6 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: ok, I've done that, cheers

File: projects/cds/public_api.ts (hunk 4 above)
Comment (bgambocjaviniar) [reply projects/cds/public_api.ts]: we can get back to this before the last merge, but we would most likely want to expose the module  and configs only for now. But it's okay for now I guess

File: projects/storefrontstyles/scss/components/cds/merchandising/carousel/_merchandising-product-carousel.scss [hunk 13]
```diff
@@ -0,0 +1,5 @@
+@import '../../../product/carousel/product-carousel';
```
Comment (bgambocjaviniar) [inline projects/storefrontstyles/scss/components/cds/merchandising/carousel/_merchandising-product-carousel.scss]: this file should be with the other carousels in the storefrontstyles lib

File: projects/storefrontstyles/scss/components/cds/_index.scss [hunk 14]
```diff
@@ -1,3 +1,4 @@
 @import '../content/index';
+@import './merchandising/carousel/merchandising-product-carousel';
```
Comment (bgambocjaviniar) [inline projects/storefrontstyles/scss/components/cds/_index.scss]: remove this file for below's comment

File: projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss (hunk 9 above)
Comment (bgambocjaviniar) [reply projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss]: Similar to what I said above, I know we have been saying cds should be isolated, but the scss should be in our styleslib

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 12 above)
Comment (bgambocjaviniar) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: my mistake (forget this :))

File: projects/storefrontstyles/scss/_components.scss (hunk 7 above)
Comment (bgambocjaviniar) [reply projects/storefrontstyles/scss/_components.scss]: Develop has it like this https://github.com/SAP/cloud-commerce-spartacus-storefront/blob/develop/projects/storefrontstyles/scss/_components.scss

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 6 above)
Comment (tobi-or-not-tobi) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: I'd argue about using `size$` going forward. A single static size property doesn't take care of different breakpoints (i.e 4 items on desktop vs 2 items on tablet vs 1 item on mobile). It also doesn't support components being (re)used in multiple grid system (i.e. using a carousel in a 3 100% width grid vs 50% width grid). 

This is where the `itemWidth` kicks in. The intend of the `itemWidth` is that it will calculate the number of items (size) for the available size vs `itemWidth`. 

If you don't agree, I'm ready to learn about the requirements.

Comment (tobi-or-not-tobi) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: Regarding `backgroundColor$` and `textColor$`, they should not be implemented in the `CarouselComponent`. They can be implemented in the `MerchandisingCarouselComponent`, as the `MerchandisingCarouselComponent` can manage those properties itself, without being bothered by the `CarouselComponent`.

File: projects/cds/src/merchandising/adapters/strategy/cds-merchandising-strategy.adapter.ts [hunk 15]
```diff
@@ -3,8 +3,8 @@ import { Injectable } from '@angular/core';
 import { ConverterService } from '@spartacus/core';
 import { Observable } from 'rxjs';
 import { CdsEndpointsService } from '../../../services/cds-endpoints.service';
+import { StrategyAdapter } from '../../connectors/strategy/cds-strategy.adapter';
 import { MERCHANDISING_PRODUCTS_NORMALIZER } from '../../connectors/strategy/converters';
-import { StrategyAdapter } from '../../connectors/strategy/strategy.adapter';
 import { MerchandisingProducts } from '../../model/merchandising.products.model';
 
 const STRATEGY_PRODUCTS_ENDPOINT_KEY = 'strategyProducts';
```
Comment (bgambocjaviniar) [inline projects/cds/src/merchandising/adapters/strategy/cds-merchandising-strategy.adapter.ts]: change the class name as well to CdsMerchandisingStrategyAdapter please

File: projects/storefrontstyles/scss/_components.scss (hunk 8 above)
Comment (bgambocjaviniar) [reply projects/storefrontstyles/scss/_components.scss]: same as above

File: projects/cds/src/merchandising/adapters/strategy/cds-merchandising-strategy.adapter.ts (hunk 15 above)
Comment (bgambocjaviniar) [reply projects/cds/src/merchandising/adapters/strategy/cds-merchandising-strategy.adapter.ts]: https://github.com/SAP/cloud-commerce-spartacus-storefront/pull/5360#pullrequestreview-315591772

from what I said here, 

the connectors/strategy should be named merchandising-strategy.adapter.ts
^ don't forget to change the class name as well please :)

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts (hunk 6 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: So, how does the existing product carousel limit the number of products returned for display. i.e. a merchandiser might want to be able to have 20 products in one carousel, but 5 in another, or 100 in another. The size relates to the number of items returned from the API endpoint. Currently this has a default limit of 16, but surely this should be configurable, no? This size property doesn't relate to the number of products to display, as such....if that makes sense

Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.ts]: I'm happy to implement the background colour/text colour in our own component only

File: projects/storefrontstyles/scss/_components.scss (hunk 7 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/_components.scss]: Its also like this in this feature branch https://github.com/SAP/cloud-commerce-spartacus-storefront/blob/feature/COL-168-cms-component/projects/storefrontstyles/scss/_components.scss, or am I missing something?

Comment (bgambocjaviniar) [reply projects/storefrontstyles/scss/_components.scss]: that link shows me this 
Screen Shot 2019-11-13 at 5 19 00 PM

They should have a space, and prettier was supposed to fix it

File: projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss (hunk 9 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/components/product/carousel/_product-carousel.scss]: This comment is on the commit and has been resolved in the latest changes

File: projects/storefrontstyles/scss/_components.scss (hunk 8 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/_components.scss]: This is a specific commit and has been resolved in a later commit

File: projects/storefrontstyles/scss/_components.scss (hunk 7 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/_components.scss]: This is a specific commit and has been resolved in a later commit

File: projects/cds/src/merchandising/adapters/strategy/cds-merchandising-strategy.adapter.ts (hunk 15 above)
Comment (darrellme) [reply projects/cds/src/merchandising/adapters/strategy/cds-merchandising-strategy.adapter.ts]: My bad, I've renamed the files AND the classes now :-)

File: projects/storefrontstyles/scss/components/cds/_index.scss (hunk 14 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/components/cds/_index.scss]: Removed the entire cds folder

File: projects/storefrontstyles/scss/components/cds/merchandising/carousel/_merchandising-product-carousel.scss (hunk 13 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/components/cds/merchandising/carousel/_merchandising-product-carousel.scss]: moved into the other carousel folder

File: projects/storefrontstyles/scss/_components.scss (hunk 7 above)
Comment (darrellme) [reply projects/storefrontstyles/scss/_components.scss]: Hmm, I've fixed it manually, but prettier didn't correct it, not did the ./build.sh sonar file. Its fixed now

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+  Pipe,
+  PipeTransform,
+  TemplateRef,
+} from '@angular/core';
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { By } from '@angular/platform-browser';
+import { RouterTestingModule } from '@angular/router/testing';
+import { Product } from '@spartacus/core';
+import { CmsComponentData } from '@spartacus/storefront';
+import { CmsMerchandisingCarouselComponent } from 'projects/cds/src/cds-models/cms.model';
```
Comment (bgambocjaviniar) [inline projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts]: please use relative path like the others

Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts]: Good catch, done

Comment (znikola) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts]: Wasn't this caught by the production CI build? Just checking myself, because if it wasn't then I missed something.

File: projects/storefrontstyles/scss/components/cds/merchandising/carousel/_merchandising-product-carousel.scss (hunk 13 above)
Comment (znikola) [reply projects/storefrontstyles/scss/components/cds/merchandising/carousel/_merchandising-product-carousel.scss]: I guess this was miscommunication on our side. Apologies for this, we should've consulted.

File: projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts (hunk 16 above)
Comment (darrellme) [reply projects/cds/src/cms-components/merchandising/carousel/merchandising-carousel/merchandising-carousel.component.spec.ts]: I don't think so as the CI build was green

File: projects/storefrontstyles/scss/components/product/carousel/_merchandising-product-carousel.scss [hunk 17]
```diff
@@ -0,0 +1,5 @@
+@import '../../../product/carousel/product-carousel';
+
+%cx-merchandising-carousel {
```
[the remark this answers is not in the document]
Comment (LTiger14) [reply projects/storefrontstyles/scss/components/product/carousel/_merchandising-product-carousel.scss]: @znikola is right these are the correct steps. Also yes the import is the correct way to do it.
